File: src/test/java/hu/numichi/reactive/logger/java/MDCContextTest.java
package hu.numichi.reactive.logger.java;

import hu.numichi.reactive.logger.MDC;
import hu.numichi.reactive.logger.exception.InvalidContextDataException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;
import reactor.util.context.Context;

import java.util.HashMap;
import java.util.Map;

import static hu.numichi.reactive.logger.Values.DEFAULT_REACTOR_CONTEXT_MDC_KEY;
import static org.junit.jupiter.api.Assertions.assertThrows;

class MDCContextTest {
    
    public static final String ANOTHER_CONTEXT_KEY = "another-context-key";
    
    @Test
    @DisplayName("should give the MDC you are looking for")
    void shouldGiveTheMDCByDefault() {
        MDC mdc1 = new MDC();
        mdc1.put("mdcKey", "mdcValue");
        
        MDC mdc2 = new MDC(ANOTHER_CONTEXT_KEY);
        mdc2.put("mdcKey", "mdcValue");
        
        
        Mono<MDC> resultDefaultByContext = Mono.deferContextual(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdc1));
        
        StepVerifier.create(resultDefaultByContext)
            .expectNext(mdc1)
            .verifyComplete();
        
        
        Mono<MDC> resultDefault = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdc1));
        
        StepVerifier.create(resultDefault)
            .expectNext(mdc1)
            .verifyComplete();
        
        
        Mono<MDC> resultAnotherByContext = Mono.deferContextual(ctx -> MDCContext.read(ctx, ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, mdc2));
        
        StepVerifier.create(resultAnotherByContext)
            .expectNext(mdc2)
            .verifyComplete();
        
        
        Mono<MDC> resultAnother = Mono.defer(() -> MDCContext.read(ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, mdc2));
        
        StepVerifier.create(resultAnother)
            .expectNext(mdc2)
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should give the MDC you are looking for (more MDC Context)")
    void shouldGiveTheMDCWithMoreMdcContext() {
        MDC defaultMdc = new MDC();
        defaultMdc.put("mdcKey", "mdcValue");
        
        MDC anotherMdc = new MDC(ANOTHER_CONTEXT_KEY);
        anotherMdc.put("mdcKey", "mdcValue");
        
        
        Mono<MDC> resultDefault = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, defaultMdc))
            .contextWrite(it -> MDCContext.put(it, anotherMdc));
        
        StepVerifier.create(resultDefault)
            .expectNext(defaultMdc)
            .verifyComplete();
        
        
        Mono<MDC> resultAnother = Mono.defer(() -> MDCContext.read(ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, defaultMdc))
            .contextWrite(it -> MDCContext.put(it, anotherMdc));
        
        StepVerifier.create(resultAnother)
            .expectNext(anotherMdc)
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should need to be able to store a map")
    void shouldNeedToBeAbleToStoreAMap() {
        Map<String, String> mdcMap = new HashMap<>();
        mdcMap.put("mdcKey", "mdcValue");
        
        
        Mono<MDC> resultDefault = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdcMap));
        
        StepVerifier.create(resultDefault)
            .expectNextMatches(mdc1 -> mdc1.asMap().equals(mdcMap) && mdc1.getContextKey().equals(DEFAULT_REACTOR_CONTEXT_MDC_KEY))
            .verifyComplete();
        
        
        Mono<MDC> resultAnother = Mono.defer(() -> MDCContext.read(ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, ANOTHER_CONTEXT_KEY, mdcMap));
        
        StepVerifier.create(resultAnother)
            .expectNextMatches(mdc1 -> mdc1.asMap().equals(mdcMap) && mdc1.getContextKey().equals(ANOTHER_CONTEXT_KEY))
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should be MDC stored with the overridden context ID")
    void shouldBeStoredWithTheOverriddenContextIDTest() {
        MDC mdc = new MDC();
        mdc.put("mdcKey", "mdcValue");
        
        Mono<MDC> result = Mono.defer(() -> MDCContext.read(ANOTHER_CONTEXT_KEY))
            .contextWrite(it -> MDCContext.put(it, ANOTHER_CONTEXT_KEY, mdc));
        
        StepVerifier.create(result)
            .expectNextMatches(mdc1 -> mdc1.asMap().equals(mdc.asMap()) && mdc1.getContextKey().equals(ANOTHER_CONTEXT_KEY))
            .verifyComplete();
    }
    
    @Test
    @DisplayName("should throw IllegalArgumentException if any parameter is NULL")
    void shouldThrowIllegalArgumentExceptionTest() {
        Map<String, String> emptyMap = new HashMap<>();
        
        assertThrows(IllegalArgumentException.class, () -> MDCContext.put(null, "", emptyMap));
        assertThrows(IllegalArgumentException.class, () -> MDCContext.put(Context.empty(), null, emptyMap));
        assertThrows(IllegalArgumentException.class, () -> MDCContext.put(Context.empty(), "", null));
    
        StepVerifier.create(MDCContext.read((String) null))
            .expectError(IllegalArgumentException.class)
            .verify();
    
        StepVerifier.create(MDCContext.read(null, ""))
            .expectError(IllegalArgumentException.class)
            .verify();
    
        StepVerifier.create(MDCContext.read(Context.of("", ""), null))
            .expectError(IllegalArgumentException.class)
            .verify();
    }
    
    @Test
    @DisplayName("should throw InvalidContextDataException if context data is invalid")
    void shouldThrowInvalidContextDataException() {
        // ANOTHER_CONTEXT_KEY is not DEFAULT_REACTOR_CONTEXT_MDC_KEY
        MDC mdc = new MDC(ANOTHER_CONTEXT_KEY);
        mdc.put("mdcKey2", "mdcValue2");
        Mono<MDC> result1 = Mono.defer(MDCContext::read)
            .contextWrite(it -> MDCContext.put(it, mdc));
        StepVerifier.create(result1)
            .expectError(InvalidContextDataException.class)
            .verify();
    
        
       
        
        
        Mono<MDC> result2 = Mono.defer(() -> MDCContext.read("not-exist-context-id"))
            .contextWrite(it -> MDCContext.put(it, mdc));
        StepVerifier.create(result2)
            .expectError(InvalidContextDataException.class)
            .verify();
        
        
        Mono<MDC> result3 = Mono.defer(MDCContext::read)
            .contextWrite(it -> it.put(DEFAULT_REACTOR_CONTEXT_MDC_KEY, ""));
        StepVerifier.create(result3)
            .expectError(InvalidContextDataException.class)
            .verify();
        
        
        Mono<MDC> result4 = Mono.defer(MDCContext::read)
            .contextWrite(it -> it.put(DEFAULT_REACTOR_CONTEXT_MDC_KEY, 100));
        StepVerifier.create(result4)
            .expectError(InvalidContextDataException.class)
            .verify();
    }
}

File: src/test/java/io/github/numichi/reactive/logger/coroutine/MDCContextTest.kt
package io.github.numichi.reactive.logger.coroutine

import io.github.numichi.reactive.logger.Configuration
import io.github.numichi.reactive.logger.DEFAULT_REACTOR_CONTEXT_MDC_KEY
import io.github.numichi.reactive.logger.MDC
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.reactor.ReactorContext
import kotlinx.coroutines.reactor.asCoroutineContext
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.withContext
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import reactor.util.context.Context

@ExperimentalCoroutinesApi
internal class MDCContextTest {

    companion object {
        const val ANOTHER_CONTEXT_KEY = "another-context-key"
    }

    @BeforeEach
    fun setUp() {
        Configuration.reset()
    }

    @Test
    fun `should give the Context with new context maps`() {
        runTest {
            val mdc1 = MDC()
            mdc1["mdcKey1"] = "mdcValue1"
            val mdc2 = MDC(ANOTHER_CONTEXT_KEY)
            mdc2["mdcKey2"] = "mdcValue2"

            val result0 = putMdc(Context.empty())
            assertEquals(0, result0.size())

            val result1 = putMdc(Context.empty(), mdc1)
            assertEquals(mdc1, result1.getOrEmpty<Map<*, *>>(Configuration.defaultReactorContextMdcKey).orElse(null))
            assertEquals(1, result1.size())

            val result2 = putMdc(Context.empty(), mdc1, mdc2)
            assertEquals(mdc1, result2.getOrEmpty<Map<*, *>>(Configuration.defaultReactorContextMdcKey).orElse(null))
            assertEquals(mdc2, result2.getOrEmpty<Map<*, *>>(ANOTHER_CONTEXT_KEY).orElse(null))
            assertEquals(2, result2.size())
        }
    }

    @Test
    fun `should give the MDC you are looking for (default key)`() {
        runTest {
            val mdc = MDC()
            mdc["mdcKey"] = "mdcValue"

            withMDCContext(mdc) {
                val result = readMdc()
                assertEquals(mdc, result)
            }

            withMDCContext(mdc) {
                val result = readMdc(coroutineContext[ReactorContext]?.context)
                assertEquals(mdc, result)
            }

            withMDCContext(Context.of(mutableMapOf("1" to "2")), mdc) {
                val size = coroutineContext[ReactorContext]?.context?.size()
                val result = readMdc(coroutineContext[ReactorContext]?.context)
                assertEquals(mdc, result)
                assertEquals(2, size)
            }
        }
    }

    @Test
    fun `should give the MDC you are looking for (default key + use native withContext)`() {
        runTest {
            val mdcMap = mutableMapOf<String, String>()
            mdcMap["mdcKey"] = "mdcValue"

            withContext(Context.of(mapOf(Configuration.defaultReactorContextMdcKey to mdcMap)).asCoroutineContext()) {
                val result = readMdc(coroutineContext[ReactorContext]?.context)
                assertEquals(mdcMap, result)
                assertEquals(mdcMap, result)
            }
        }
    }

    @Test
    fun `should give the MDC you are looking for (another key)`() {
        runTest {
            val mdc2 = MDC(ANOTHER_CONTEXT_KEY)
            mdc2["mdcKey"] = "mdcValue"

            withMDCContext(mdc2) {
                val result = readMdc(ANOTHER_CONTEXT_KEY)
                assertEquals(mdc2, result)
            }
        }
    }

    @Test
    fun `should give the MDC you are looking for (more MDC Context)`() {
        runTest {
            val mdc1 = MDC()
            mdc1["mdcKey"] = "mdcValue"

            val mdc2 = MDC(ANOTHER_CONTEXT_KEY)
            mdc2["mdcKey"] = "mdcValue"

            withMDCContext(mdc1, mdc2) {
                val result1 = readMdc(coroutineContext[ReactorContext]?.context)
                val result2 = readMdc(coroutineContext[ReactorContext]?.context, Configuration.defaultReactorContextMdcKey)
                val result3 = readMdc(Configuration.defaultReactorContextMdcKey)
                val result4 = readMdc(coroutineContext[ReactorContext]?.context, ANOTHER_CONTEXT_KEY)
                val result5 = readMdc(ANOTHER_CONTEXT_KEY)

                assertEquals(mdc1, result1)
                assertEquals(mdc1, result2)
                assertEquals(mdc1, result3)
                assertEquals(mdc2, result4)
                assertEquals(mdc2, result5)
            }
        }
    }

    @Test
    fun `should need to be able to store a map`() {
        runTest {
            val mdcMap = mutableMapOf<String, String>()
            mdcMap["mdcKey"] = "mdcValue"

            withMDCContext(MDC(mdcMap)) {
                val result1 = readMdc(coroutineContext[ReactorContext]?.context)
                val result2 = readMdc()
                assertEquals(mdcMap, result1)
                assertEquals(mdcMap, result2)
            }

            withMDCContext(MDC(ANOTHER_CONTEXT_KEY, mdcMap)) {
                val result = readMdc(coroutineContext[ReactorContext]?.context, ANOTHER_CONTEXT_KEY)
                assertEquals(mdcMap, result)
            }
        }
    }

    @Test
    fun `should be MDC stored with the overridden context ID`() {
        runTest {
            val mdcMap = mutableMapOf<String, String>()
            mdcMap["mdcKey"] = "mdcValue"

            withMDCContext(MDC(mdcMap)) {
                val result1 = readMdc(coroutineContext[ReactorContext]?.context)
                assertEquals(mdcMap, result1)
            }

            withMDCContext(MDC(ANOTHER_CONTEXT_KEY, mdcMap)) {
                val result1 = readMdc(coroutineContext[ReactorContext]?.context, ANOTHER_CONTEXT_KEY)
                assertEquals(mdcMap, result1)
            }
        }
    }

    @Test
    fun `should throw IllegalArgumentException if any parameter is NULL`() {
        runTest {
            assertThrows<IllegalArgumentException> { withMDCContext(null, MDC()) {} }
            assertThrows<IllegalArgumentException> { readMdc(null) }
            assertThrows<IllegalArgumentException> { readMdc(null, "any-key") }
        }
    }

    @Test
    fun `should run toString method if value not String`() {
        runTest {
            val ctx = Context.of(mapOf(
                DEFAULT_REACTOR_CONTEXT_MDC_KEY to mapOf(
                    "int" to 111,
                    "map" to mapOf<Any, Any>()
                )
            ))

            withContext(ctx.asCoroutineContext()) {
                val result2 = readMdc()
                assertEquals(2, result2.size)
                assertEquals("111", result2["int"])
                assertEquals("{}", result2["map"])
            }
        }
    }

    @Test
    fun `should read empty MDC when not found contextKey`() {
        runTest {
            withContext(Context.empty().asCoroutineContext()) {
                val result2 = readMdc()
                assertEquals(0, result2.size)
            }
        }
    }
}